File: scripts/metrics/example_metrics_run.m
% scripts/metrics/example_metrics_run.m
% Generates synthetic transition data and computes metric table.

fprintf('== example_metrics_run ==\n');
fprintf('Timestamp: %s\n', datestr(now, 31));

% Synthetic profile: hover -> cruise transition over ~20 s.
dt = 0.05;
t = (0:dt:30)';
N = numel(t);

sig = @(x) 1 ./ (1 + exp(-x));
transitionCore = sig((t-10)/1.5);

tilt_deg = 90 * transitionCore;
speed = 55 * transitionCore + 2.5 * exp(-0.15*(t-12).^2); % slight overshoot bump

accel_x = [diff(speed)./dt; 0];
accel_xyz = [accel_x, zeros(N,1), 0.05*sin(0.8*t)];
jerk_xyz = [[diff(accel_x)./dt; 0], zeros(N,1), [diff(accel_xyz(:,3))./dt; 0]];

% 12 pseudo actuator channels (normalized command)
u_front = repmat(0.3 + 0.5*transitionCore, 1, 6);
u_rear  = repmat(0.7 - 0.4*transitionCore, 1, 6);
control = [u_front, u_rear];

logData = struct();
logData.t = t;
logData.speed = speed;
logData.tilt_deg = tilt_deg;
logData.accel_xyz = accel_xyz;
logData.jerk_xyz = jerk_xyz;
logData.control = control;

metrics = compute_transition_metrics(logData, struct('target_speed', 55));

metricNames = fieldnames(metrics);
metricVals = struct2cell(metrics);
T = table(metricNames, metricVals, 'VariableNames', {'Metric', 'Value'});

disp(T);

outDir = fullfile('docs', 'evidence');
if ~exist(outDir, 'dir'); mkdir(outDir); end

save(fullfile(outDir, 'metrics_example_latest.mat'), 'metrics', 'T');

try
    writetable(T, fullfile(outDir, 'metrics_example_latest.csv'));
catch
    % If writetable unavailable, skip CSV export.
end

fprintf('METRICS_EXAMPLE: PASS\n');

File: scripts/metrics/example_metrics_run.py
#!/usr/bin/env python3
"""Generate an example transition metrics table (Python helper)."""

from __future__ import annotations

from datetime import datetime
import json
from pathlib import Path
import sys

import numpy as np

REPO_ROOT = Path(__file__).resolve().parents[2]
if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))

from scripts.metrics.compute_transition_metrics import compute_transition_metrics


def main():
    dt = 0.05
    t = np.arange(0.0, 30.0 + 1e-12, dt)
    sig = lambda x: 1.0 / (1.0 + np.exp(-x))
    core = sig((t - 10.0) / 1.5)

    tilt = 90.0 * core
    speed = 55.0 * core + 2.5 * np.exp(-0.15 * (t - 12.0) ** 2)

    accel_x = np.gradient(speed, dt)
    accel = np.column_stack([accel_x, np.zeros_like(t), 0.05 * np.sin(0.8 * t)])
    jerk = np.gradient(accel, dt, axis=0)

    u_front = np.tile(0.3 + 0.5 * core, (6, 1)).T
    u_rear = np.tile(0.7 - 0.4 * core, (6, 1)).T
    control = np.hstack([u_front, u_rear])

    log = {
        "t": t,
        "speed": speed,
        "tilt_deg": tilt,
        "accel_xyz": accel,
        "jerk_xyz": jerk,
        "control": control,
    }

    metrics = compute_transition_metrics(log, target_speed=55.0)

    out = {
        "generated_at": datetime.now().isoformat(),
        "metrics": metrics,
    }

    repo = Path(__file__).resolve().parents[2]
    out_json = repo / "docs" / "evidence" / "metrics_example_latest.json"
    out_json.parent.mkdir(parents=True, exist_ok=True)
    out_json.write_text(json.dumps(out, indent=2))

    print(f"Wrote {out_json}")
    for k, v in metrics.items():
        print(f"{k:24s}: {v:.6f}" if isinstance(v, (float, int)) else f"{k:24s}: {v}")


if __name__ == "__main__":
    main()
